### ISSUE TITLE
DCamProf profiles: Neutral tone reproduction operator not applied correctly

### ISSUE THREAD

**Issue by stefan-i** [Beep6581/RawTherapee#3697]
I noticed a problem with the handling of DCP profiles created with DCamProf which contain a tone curve that makes use of DCamProf's neutral tone reproduction operator (NTRO). 

**Problem**: 
The DCP tone curve is rendered simply as "film like" and not with the NTRO.

**How to reproduce**:

1. Load Raw file and make sure a DCamProf profile is used in color management tab
2. Disable "DCP tone curve" in color management
3. In exposure tab, load the same tone curve that was used to create the profile with DCamProf and choose "film-like" rendering
Result: The rendering of 1) and 3) are exactly the same

If the DCP tone curve was handled correctly we would expect to see a clear difference in the rendering e.g. of saturated red colors (these should be less bright and saturated with the NTRO, similar to the "perceptual" tone curve mode).

The problem seems to occur with all DCamProf profiles, those shipped with RT and those I created myself. I am pretty sure that the DCP tone curve handling was correct in previous RT versions. I did similar tests shortly after I created my DCPs about a year ago and IIRC the NTRO worked as expected back then. These are the instructions and commandline I followed to create my profiles (copied straight from DCamProf manual):

```
Convert the native profile to a DCP, and apply a tone curve using DCamProf’s neutral 
tone reproduction operator that strives to maintain a neutral and realistic color 
appearance despite the curve:

 dcamprof make-dcp -n "Camera Name" -d "My Profile" \
   -t acr -g adobergb-strong my-profile.json my-profile.dcp
```
 
If I am correct this is a serious issue because all the DCPs created with DCamProf that are shipped with RT do not work as intended and lead to wrong colors because the NTRO is not applied.
These are the versions I used to reproduce the problem:

Version: 5.0-r1-gtk3
Branch: 5.0-r1-gtk3
Commit: 50114c1
Commit date: 2017-02-01
Compiler: gcc 5.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.18.0
Build type: release
Build flags: -m64 -mwin32 -mthreads -std=gnu++11 -mtune=generic -Werror=unused-label -fopenmp -Werror=unknown-pragmas -msse -msse2 -mwindows -fopenmp -Wno-aggressive-loop-optimizations -DNDEBUG -O3
Link flags: -m64 -mwin32 -mthreads -static-libgcc -mtune=generic -mwindows -s -O3
OpenMP support: ON
MMAP support: ON

Branch: gtk3
Version: 4.2.1234
Changeset: d16d244fc572615248a8c0f741f434213f4cd548
Compiler: gcc 5.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.18.0
Build type: release
Build flags: -m64 -mwin32 -mthreads -std=gnu++11 -mtune=generic -Werror=unused-label -fopenmp -Werror=unknown-pragmas -msse -msse2 -mwindows -fopenmp -Wno-aggressive-loop-optimizations -DNDEBUG -O3
Link flags: -m64 -mwin32 -mthreads -static-libgcc -mtune=generic -mwindows -s -O3
OpenMP support: ON
MMAP support: ON

So the wrong tone curve handling probably was introduced some time before 4.2.1234

**Comment by Floessie** [comment]
@heckflosse Maybe we broke it in #3343?

**Comment by heckflosse** [comment]
@Floessie I don't think so. But I will check.

**Comment by heckflosse** [comment]
@atorger @Beep6581 @iliasg know more about dcp

**Comment by heckflosse** [comment]
[From here](https://www.ludd.ltu.se/~torger/dcamprof.html#tro_dcp_notes)

```
When making a DNG profile the operator is implemented through the LookTable and curve. So if you strip away the LookTable and curve you have the pure colorimetric profile left.
```
Doesn't that mean you have to enable the `Look table` to make use of DCamProf's neutral tone reproduction operator?

But I'm not a specialist in dcp.

**Comment by stefan-i** [comment]
Yes, the look table (and of course base table) is always enabled in my examples, that's not the issue.

**Comment by heckflosse** [comment]
Ok, that wasn't clear in first post. Then we'll have to wait for answers of specialists.

**Comment by Beep6581** [comment]
> If I am correct this is a serious issue because all the DCPs created with DCamProf that are shipped with RT do not work as intended and lead to wrong colors because the NTRO is not applied.

I agree.

Tested using RawTherapee `5.0-r1-gtk3-23-g5696df39` and DCamProf `1.0.5`.
Raw file: http://rawtherapee.com/shared/test_images/colortarget/PENTAX%20K10D%20daylight.dng

Neutral + DCP with only the base table (HueSatMap) applied:
![imgur-2017_02_17-18 49 52](https://cloud.githubusercontent.com/assets/5844619/23078828/8daa9960-f54a-11e6-84e2-5fc43f810529.png)

Enable the tone curve (`acr`) and look table (default `ntro_conf.json`):
![imgur-2017_02_17-18 50 11](https://cloud.githubusercontent.com/assets/5844619/23078829/8dab9fae-f54a-11e6-8101-72992b65be4a.png)

Disable DCP tone curve and look table, enable spline tone curve using the same "acr" shape [acr.rtc.txt](https://github.com/Beep6581/RawTherapee/files/784064/acr.rtc.txt) in Exposure tool using Film-like method:
![imgur-2017_02_17-18 50 30](https://cloud.githubusercontent.com/assets/5844619/23078827/8da9d26e-f54a-11e6-9a68-ded0cbb9db52.png)

Enable look table:
![imgur-2017_02_17-18 52 17](https://cloud.githubusercontent.com/assets/5844619/23078830/8db40ee6-f54a-11e6-9fb1-61ae86437d1f.png)

> Problem:
The DCP tone curve is rendered simply as "film like" and not with the NTRO.

That's one of several possible conclusions. I observe that the film-like curve matches the result of the DCP curve only when the look table is enabled.

* One conclusion could be the one you gave.
* another could be that the film-like curve behaves similarly to the way the DCP curve by itself behaves,
* or coincidence.

I edited the neutral tone reproduction operator to desaturate all colors [ntro3_conf.json.txt](https://github.com/Beep6581/RawTherapee/files/784142/ntro3_conf.json.txt), then created a new DCP [k10d_acr_ntro3.zip](https://github.com/Beep6581/RawTherapee/files/784156/k10d_acr_ntro3.zip) using it.

Using an "acr" exposure curve, film-like, with DCP look table enabled:
![imgur-2017_02_17-20 46 38](https://cloud.githubusercontent.com/assets/5844619/23080843/a7bb5724-f552-11e6-82d9-5ff90fbf0ce3.png)

Exposure curve neutral, DCP curve enabled:
![imgur-2017_02_17-20 48 37](https://cloud.githubusercontent.com/assets/5844619/23080902/e569a4f4-f552-11e6-9cde-a7399d021f27.png)

So it seems that the neutral tone reproduction operator is indeed used.

**Comment by Beep6581** [comment]
If this is still valid, then it seems to confirm my conclusion:
https://github.com/Beep6581/RawTherapee/issues/3133#issuecomment-179851251

> DNG profiles are applied in two stages, first one scene-referred pass with huesatmap which is before exposure correction, and then a second output-referred pass with the looktable and curve, which should be after exposure correction. In older versions of RT everything was applied in the first pass, but that's many versions ago.
The DCP handling should be made according to the DNG specification.

Note that "Film-like" means "Adobe-like".

It seems from the code that DCP stuff is applied in the pipeline after the exposure compensation slider, but before the curves.
https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/improcfun.cc#L3447

**Comment by stefan-i** [comment]
Thank you for your investigations. But where does this leave us? Is it working as expected or is there possibly something wrong? 
I am still confused about the film-like rendering of the embedded DCP curve. Or is this the only way those curves can be rendered according to the DNG specification and all the NTRO magic is done via the Look Table?

**Comment by Beep6581** [comment]
@stefan-i I believe the NTRO works fine.
I defer to @atorger.

**Comment by atorger** [comment]
I haven't read this full thread. Just a comment on DCP/Curves/NTRO. I'm quite sure that the DNG specification does not tell what type of operator that should be used on the tone curve, but defacto standard is Adobe's "film-like", published via the DNG SDK. When things are unclear in the spec, the DNG SDK serves as a reference.

When coding DCP processing for RT the guiding principle is to make it as "standard" as possible, which makes it possible to use DCPs generated by various profile makers and interchange between raw converters.

DCamProf profiles can be used in Adobe Camera Raw too, and thus the "NTRO magic" happens in the looktable, which expects that a Adobe Film-Like curve is used for the tonecurve. If some other curve-type would be used the profile would not show colors as intended.

**Comment by stefan-i** [comment]
Thank you @atorger for your explanation. I am glad to hear everything is working fine. So is it correct to say that the Look table is primarily meant to "repair" the color shifts introduced by the film-like DCP tone curve?
I am also wondering what would be the best way of further manipulating an image in RT with tone curves while keeping further color shifts at a minimum. Is it best to stay with film-like tone curves as the look table is "tuned" for their rendering, or would the "perceptual" rendering be more suitable because the look table is tuned only for the specific curve shape embedded in the DCP?
